### PR TITLE
Closes #481: Fixing referenceextraction/softwareurl/output_transformer

### LINKED ISSUE #481: referenceextraction/softwareurl/output_transformer integration test fails
```
2016-06-13 12:03:56,248 WARN [main] org.apache.hadoop.mapred.YarnChild: Exception running child : org.apache.pig.backend.executionengine.ExecException: ERROR 0: Exception while executing [POUserFunc (Name: POUserFunc(eu.dnetlib.iis.common.pig.udfs.DeduplicateIdsWithConfidence)[bag] - scope-8 Operator Key: scope-8) children: null at []]: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Float
	at org.apache.pig.backend.hadoop.executionengine.physicalLayer.PhysicalOperator.getNext(PhysicalOperator.java:338)
	at org.apache.pig.backend.hadoop.executionengine.physicalLayer.relationalOperators.POForEach.processPlan(POForEach.java:378)
	at org.apache.pig.backend.hadoop.executionengine.physicalLayer.relationalOperators.POForEach.getNextTuple(POForEach.java:298)
	at org.apache.pig.backend.hadoop.executionengine.mapReduceLayer.PigGenericMapReduce$Reduce.runPipeline(PigGenericMapReduce.java:464)
	at org.apache.pig.backend.hadoop.executionengine.mapReduceLayer.PigGenericMapReduce$Reduce.processOnePackageOutput(PigGenericMapReduce.java:432)
	at org.apache.pig.backend.hadoop.executionengine.mapReduceLayer.PigGenericMapReduce$Reduce.reduce(PigGenericMapReduce.java:412)
	at org.apache.pig.backend.hadoop.executionengine.mapReduceLayer.PigGenericMapReduce$Reduce.reduce(PigGenericMapReduce.java:256)
	at org.apache.hadoop.mapreduce.Reducer.run(Reducer.java:171)
	at org.apache.hadoop.mapred.ReduceTask.runNewReducer(ReduceTask.java:627)
	at org.apache.hadoop.mapred.ReduceTask.run(ReduceTask.java:389)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:163)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1671)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:158)
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Float
	at eu.dnetlib.iis.common.pig.udfs.DeduplicateIdsWithConfidence.exec(DeduplicateIdsWithConfidence.java:44)
	at eu.dnetlib.iis.common.pig.udfs.DeduplicateIdsWithConfidence.exec(DeduplicateIdsWithConfidence.java:22)
	at org.apache.pig.backend.hadoop.executionengine.physicalLayer.expressionOperators.POUserFunc.getNext(POUserFunc.java:330)
	at org.apache.pig.backend.hadoop.executionengine.physicalLayer.expressionOperators.POUserFunc.getNextDataBag(POUserFunc.java:374)
	at org.apache.pig.backend.hadoop.executionengine.physicalLayer.PhysicalOperator.getNext(PhysicalOperator.java:309)
	... 14 more
```

### PR DESCRIPTION
**Pull request by marekhorst**
…integration test by parameterizing `DeduplicateIdsWithConfidence` UDF with `confidenceLevelPosition`.

### CODE REVIEW

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java [hunk 1]
```diff
@@ -13,55 +13,76 @@
...
+
+    public DeduplicateIdsWithConfidence() {
+        this(1);
+    }
+    
+    public DeduplicateIdsWithConfidence(int confidenceLevelPosition) {
+        this.confidenceLevelPosition = confidenceLevelPosition;
+    }
+    
+    public DeduplicateIdsWithConfidence(String confidenceLevelPosition) {
```
Comment (lukdumi) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: It's very strange. If you pass a string that is not an integer, you'll get exception, so why do you allow it?

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java [hunk 2]
```diff
@@ -13,55 +13,76 @@
...
+        this(Integer.valueOf(confidenceLevelPosition));
+    }
+
+    //------------------------ PUBLIC --------------------------
+    
     @Override
     public DataBag exec(Tuple tuple) throws IOException {
```
Comment (lukdumi) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: No comment here, above the method.

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java [hunk 3]
```diff
@@ -13,55 +13,76 @@
...
- * Highest confidence level is picked when identifier duplicate is found.
- * Identifiers are sorted lexicographically.
+ * Deduplicates bag of tuples where tuple[0] is identifier and tuple[confidenceLevelPosition] is confidence level. 
+ * Highest confidence level is picked when identifier duplicate is found. Identifiers are sorted lexicographically.
  *
  * @author mhorst
  */
 public class DeduplicateIdsWithConfidence extends EvalFunc<DataBag> {
 
+    /**
```
Comment (lukdumi) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: I don't understand why you start a name of the class with a verb (it resembles the method name). Should it be named like for example IdWithConfidenceExtractor?

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java [hunk 4]
```diff
@@ -13,55 +13,76 @@
...
+                                dedupMap.put((String) currentTuple.get(0), currentTuple);
+                            }
+                        } else {
+                            dedupMap.put((String) currentTuple.get(0), currentTuple);
+                        }
+                    }
+                } else {
+                    dedupMap.put((String) currentTuple.get(0), currentTuple);
+                }
             }
```
Comment (lukdumi) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: Marek, this code is a bit of confusing to read. 
Firstly, variables names do not say much, the names inform about types of variables, but it's obvious.
Secondly, I suggest to divide code into blocks of private methods with relevant names saying what they do. For example you could put the lines 60-71 into a method 'chooseTupleWithHigherConfLevel(currentTuple, storedTuple)' that would return the tuple with the higher confidenceLevel. This way the code would be much easier to read and maintain.

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java (hunk 1 above)
Comment (marekhorst) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: This is 'fun with PIG'. Appropriate comment added.

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java (hunk 3 above)
Comment (marekhorst) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: I don't recall the reasoning behind this, but I agree, this might be confusing.

What it does is whole tuples deduplication by id and highest confidence level (so the original deduplicated tuples are returned). Now, what name to pick not to be too wordy... `IdentifiedTupleByConfidenceDeduplicator`?

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java (hunk 4 above)
Comment (marekhorst) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: I'd go with `updateStoredTupleWhenConfidenceLevelHigher(currentTuple, deduplicatedTuplesMap)` since when new candidate is not better than stored one we wouldn't need to perform any operation on map. I've also simplified some of the conditions.

As for the names - I fixed them where I could but generally we are dealing with tuples and we do not know what they are so it's difficult to name them appropriately. They could be anything with `identifier` as first field and `confidenceLevel` among other fields.

**Comment by marekhorst** [comment]
I've just pushed my changes except renaming `IdWithConfidenceExtractor`. Let's agree on the name (see my candidate) then I will push another commit.

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java (hunk 3 above)
Comment (lukdumi) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: Well, what about simply IdConfidenceTupleDeduplicator? You have written in comments how it is actually deduplicated.

File: iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java (hunk 4 above)
Comment (lukdumi) [inline iis-common/src/main/java/eu/dnetlib/iis/common/pig/udfs/DeduplicateIdsWithConfidence.java]: OK, now it is much easier to read :+1: 

As for the names, yes, I agree, it is a sort of generic code, but things like _inputTuplesIterator_ are much more readable than _it_ ;-)

**Comment by lukdumi** [comment]
Marek, it quite good now. Only the name of the class has remained.

**Comment by marekhorst** [comment]
Name was fixed.
